src/lib/operators/abstract_operator.cpp

```diff
@@ -24,20 +24,55 @@ namespace opossum {
 AbstractOperator::AbstractOperator(const OperatorType type, const std::shared_ptr<const AbstractOperator>& left,
                                    const std::shared_ptr<const AbstractOperator>& right,
                                    std::unique_ptr<AbstractOperatorPerformanceData> init_performance_data)
-    : performance_data(std::move(init_performance_data)), _type(type), _left_input(left), _right_input(right) {}
+    : performance_data(std::move(init_performance_data)), _type(type), _left_input(left), _right_input(right) {
+  // Tell input operators that we want to consume their output
+  if (_left_input) mutable_left_input()->register_consumer();
+  if (_right_input) mutable_right_input()->register_consumer();
+}
+
+AbstractOperator::~AbstractOperator() {
+  if constexpr (HYRISE_DEBUG) {
+    bool left_has_executed = _left_input ? _left_input->performance_data->executed : false;
+    bool right_has_executed = _right_input ? _right_input->performance_data->executed : false;
+
+    auto transaction_context = _transaction_context.has_value() ? _transaction_context->lock() : nullptr;
+    bool aborted = transaction_context ? transaction_context->aborted() : false;
+
+    /**
+     * Hack: The condition _consumer_count == 0 is useful for some tests that create operators, but do not execute
+     *       them. We do not want to force tests to call execute() on operators when their only purpose is to test the
+     *       output of, for example, the description() function.
+     */
+    Assert(this->performance_data->executed || aborted || !left_has_executed || !right_has_executed ||
+               _consumer_count == 0,
+           "Operator did not execute, but at least one input operator has.");
+  }
+}
 
 OperatorType AbstractOperator::type() const { return _type; }
 
 void AbstractOperator::execute() {
   DTRACE_PROBE1(HYRISE, OPERATOR_STARTED, name().c_str());
-  DebugAssert(!_left_input || _left_input->get_output(), "Left input has not yet been executed");
-  DebugAssert(!_right_input || _right_input->get_output(), "Right input has not yet been executed");
-  DebugAssert(!performance_data->executed, "Operator has already been executed");
+
+  /**
+   * If an operator has already executed, we return without re-execution in the following two cases:
+   *    a) _output is set                           ->  Reuse results
+   *    b) _output got cleared && zero consumers    ->  Operator results are no more relevant. Probably, a subquery
+   *                                                    has already executed the operator and all of its successors.
+   */
+  if (performance_data->executed && (_output || _consumer_count == 0)) return;
```

Comment:
Following scenario: 
* Operator A becomes scheduled, but operator B is executed before A.  
* Operator B contains a subquery that needs the results of A. Since there are no results available, the ExpressionEvaluator executes operator A. 
* After some time, the scheduler wants to execute A, but the results are already there. Therefore, we can return immediately, preventing a re-execution.

Reply:
And here is another, more advanced scenario:

<img width="878" alt="image" src="https://user-images.githubusercontent.com/9380917/105994526-03d51300-60a8-11eb-9fc2-12a55bff278a.png">

Reply:
Changed to
```cpp
  if (_executed) return;
  Assert((_consumer_count > 0 || _output), "Unexpected execution of an operator.");
```